pkg/kubelet/cm/deviceplugin/manager.go

```diff
@@ -446,155 +413,86 @@ func (m *ManagerImpl) readCheckpoint() error {
-	}
-	allocated := available.UnsortedList()[:needed]
-	// Updates m.allocatedDevices with allocated devices to prevent them
-	// from being allocated to other pods/containers, given that we are
-	// not holding lock during the rpc call.
-	for _, device := range allocated {
-		m.allocatedDevices[resource].Insert(device)
-		devices.Insert(device)
-	}
-	return devices, nil
-}
 
-// allocateContainerResources attempts to allocate all of required device
-// plugin resources for the input container, issues an Allocate rpc request
-// for each new device resource requirement, processes their AllocateResponses,
-// and updates the cached containerDevices on success.
-func (m *ManagerImpl) allocateContainerResources(pod *v1.Pod, container *v1.Container) error {
-	podUID := string(pod.UID)
-	contName := container.Name
-	allocatedDevicesUpdated := false
-	for k, v := range container.Resources.Limits {
-		resource := string(k)
-		needed := int(v.Value())
-		glog.V(3).Infof("needs %d %s", needed, resource)
-		_, registeredResource := m.allDevices[resource]
-		_, allocatedResource := m.allocatedDevices[resource]
-		// Continues if this is neither an active device plugin resource nor
-		// a resource we have previously allocated.
-		if !registeredResource && !allocatedResource {
-			continue
-		}
-		// Updates allocatedDevices to garbage collect any stranded resources
-		// before doing the device plugin allocation.
-		if !allocatedDevicesUpdated {
-			m.updateAllocatedDevices(m.activePods())
-			allocatedDevicesUpdated = true
-		}
-		allocDevices, err := m.devicesToAllocate(podUID, contName, resource, needed)
-		if err != nil {
-			return err
-		}
-		if allocDevices == nil || len(allocDevices) <= 0 {
-			continue
-		}
-		startRPCTime := time.Now()
-		// devicePluginManager.Allocate involves RPC calls to device plugin, which
-		// could be heavy-weight. Therefore we want to perform this operation outside
-		// mutex lock. Note if Allocate call fails, we may leave container resources
-		// partially allocated for the failed container. We rely on updateAllocatedDevices()
-		// to garbage collect these resources later. Another side effect is that if
-		// we have X resource A and Y resource B in total, and two containers, container1
-		// and container2 both require X resource A and Y resource B. Both allocation
-		// requests may fail if we serve them in mixed order.
-		// TODO: may revisit this part later if we see inefficient resource allocation
-		// in real use as the result of this. Should also consider to parallize device
-		// plugin Allocate grpc calls if it becomes common that a container may require
-		// resources from multiple device plugins.
-		m.mutex.Lock()
-		e, ok := m.endpoints[resource]
-		m.mutex.Unlock()
+	pod := attrs.Pod
+	var err error
+	request, _ := m.getPodRequests(pod)
+	selected, err := m.allocation.reserve(m.allDevices, request)
```

Comment:
I'm also wondering if you don't have a race here too with `genericDeviceUpdateCallback` changing the list underneath you

Reply:
Good question. In your PR #52616, the device registration logic has been reworked and is abstracted in a better way. The picture in my mind is like this: the endpoint handler will handle device registration / reregistration so the manager can always rely on that interface to get a list of "all devices". This PR focuses on the allocation / reallocation logic so the "allocated devices" is properly taken care of.

When PR #52616 is in, we will need to revisit the workflow so that the allocation logic postpones its scanning of "all devices" to a late point in time, *by invoking the endpoint interface function*.